### PR TITLE
Add ,shuffle option

### PR DESCRIPTION
**Pull request by taahamahdi**

### CODE REVIEW

File: GAMEPLAY.md [hunk 1]
```diff
@@ -54,12 +54,15 @@ Setting the goal ends the game when the given goal score is reached. For example
...
-
 ## ,timer [time (seconds)]
 Setting a timer limits players to guess in under `time` seconds before the round ends automatically. Once a user gives a valid timeout, such as `,timer 10`, the timer will automatically start at the beginning of every round. If no one guesses right in the time allotted (in this example, 10 seconds), the round is over. **Typing** `,timer` **disables the timer**. Set the timer under 10 (5? 3?) seconds and face off with your friends to see who the ultimate KMQ champ is! Alternatively, set it above 30 seconds, and enjoy a variety of your favorite artists while expanding your music library.
 
+![timer](/images/timer.png)
+
+## ,shuffle [random|unique]
+Setting the shuffle mode to unique limits each random song chosen to appear at most once per game. Conversely, `,shuffle random` picks a song at random from the given options every round, so songs may play multiple times in a game. To limit instances where songs would randomly (with a low statistical probability) play multiple times back-to-back, no song will repeat in the span of 10 songs using `,shuffle random`.
```
Comment (Brainicism) [inline GAMEPLAY.md:63]: The wording here is too technical

File: src/models/game_session.ts [hunk 2]
```diff
@@ -185,10 +185,20 @@ export default class GameSession {
             return;
         }
 
+        if (guildPreference.getShuffleType() === ShuffleType.UNIQUE && guildPreference.getLimit() === this.lastPlayedSongsQueue.length) {
+            logger.info(`${getDebugContext(message)} | Game session ended (all ${guildPreference.getLimit()} unique songs played)`);
+            await sendEndGameMessage({ channel: message.channel, authorId: message.author.id }, this);
+            await this.endSession();
```
Comment (Brainicism) [inline src/models/game_session.ts:191]: Should the game end once all songs have been played? Or restart the queue

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/models/game_session.ts (hunk 2 above)
Comment (taahamahdi) [reply src/models/game_session.ts:191]: Either one works, which do you prefer?

Comment (taahamahdi) [reply src/models/game_session.ts:191]: We could play a different sound or send a message when all songs have played and the queue is resetting

File: GAMEPLAY.md [hunk 3]
```diff
@@ -54,12 +54,15 @@ Setting the goal ends the game when the given goal score is reached. For example
...
-
 ## ,timer [time (seconds)]
 Setting a timer limits players to guess in under `time` seconds before the round ends automatically. Once a user gives a valid timeout, such as `,timer 10`, the timer will automatically start at the beginning of every round. If no one guesses right in the time allotted (in this example, 10 seconds), the round is over. **Typing** `,timer` **disables the timer**. Set the timer under 10 (5? 3?) seconds and face off with your friends to see who the ultimate KMQ champ is! Alternatively, set it above 30 seconds, and enjoy a variety of your favorite artists while expanding your music library.
 
+![timer](/images/timer.png)
+
+## ,shuffle [random|unique]
+Setting the shuffle mode to unique plays through every song once before any are repeated. With `,shuffle random`, songs matching the user's criteria are randomly chosen from before every round, so there is a chance of repeated songs within a timespan, depending on the total amount of songs.
```
Comment (Brainicism) [inline GAMEPLAY.md:63]: `unique` should be in code blocks like `random`

Comment (Brainicism) [reply GAMEPLAY.md:63]: Remove the last comma

File: src/commands/game_options/shuffle.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+export enum ShuffleType {
+    RANDOM = "random",
+    UNIQUE = "unique",
+}
+
+export default class ShuffleCommand implements BaseCommand {
+    async call({ message, parsedMessage }: CommandArgs) {
+        const guildPreference = await getGuildPreference(message.guildID);
+        const shuffleType = parsedMessage.components[0].toLowerCase();
```
Comment (Brainicism) [inline src/commands/game_options/shuffle.ts:17]: Cast to `ShuffleType` here

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/models/game_session.ts [hunk 5]
```diff
@@ -185,10 +185,17 @@ export default class GameSession {
...
+            this.lastPlayedSongsQueue = [];
+        } else if (guildPreference.getShuffleType() === ShuffleType.RANDOM && this.lastPlayedSongsQueue.length >= LAST_PLAYED_SONG_QUEUE_SIZE) {
+            this.lastPlayedSongsQueue.shift();
+        }
+
         this.sessionInitialized = true;
         let randomSong: QueriedSong;
         try {
-            if (guildPreference.getLimit() <= LAST_PLAYED_SONG_QUEUE_SIZE) {
+            if (guildPreference.getShuffleType() === ShuffleType.RANDOM && guildPreference.getLimit() <= LAST_PLAYED_SONG_QUEUE_SIZE) {
```
Comment (Brainicism) [inline src/models/game_session.ts:198]: Move this logic into line 191

**Review by Brainicism — CHANGES_REQUESTED** [review]

**Review by Brainicism — DISMISSED** [review]

**Review by Brainicism — APPROVED** [review]

**Pull request merged**